build

```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+set -e
+
+# change to this script's directory
+# credits to https://stackoverflow.com/questions/4774054/reliable-way-for-a-bash-script-to-get-the-full-path-to-itself/12197518#12197518
+pushd . > /dev/null
+SCRIPT_PATH="${BASH_SOURCE[0]}";
+while([ -h "${SCRIPT_PATH}" ]); do
+    cd "`dirname "${SCRIPT_PATH}"`"
+    SCRIPT_PATH="$(readlink "`basename "${SCRIPT_PATH}"`")";
+done
+cd "`dirname "${SCRIPT_PATH}"`" > /dev/null
+SCRIPT_PATH="`pwd`";
+popd  > /dev/null
+echo "change working directory to '$SCRIPT_PATH'"
+cd "$SCRIPT_PATH"
+
+set -x
+
+cargo clippy --all-targets --all-features -- -D warnings
+# NOTE: for --release builds also change ./run to actually use it!
+cargo build
+cargo check
```

Comment:
@flip1995 `cargo build` already performs a `cargo check`, right? And doesn't `cargo clippy` performs a `cargo build`? so are these two compiler invocations necessary?

Reply:
`cargo build` performs `cargo check`. `cargo clippy` builds all dependencies, but only checks the crate IIRC.

Yes looks like it:
https://github.com/rust-lang-nursery/rust-clippy/blob/6ae89c4f1123d81e56141b02cd3b83436c1ad035/src/driver.rs#L88-L92

Reply:
But theoretically 
```
cargo clippy
cargo test
```
should be enough

Reply:
I think `check` is in fact covered by `build` (or `clippy`), but `clippy` doesn't seem to invoke `build`:
```
 $ cargo clean
 $ ./build
change working directory to '/home/joshua/repos/by-owner/comprakt/comprakt'
+ cargo clippy --all-targets --all-features -- -D warnings
   Compiling proc-macro2 v0.4.20
   Compiling unicode-xid v0.1.0
    Checking libc v0.2.43
    Checking unicode-width v0.1.5
    Checking vec_map v0.8.1
    Checking ansi_term v0.11.0
    Checking strsim v0.7.0
    Checking bitflags v1.0.4
    Checking textwrap v0.10.0
    Checking atty v0.2.11
    Checking clap v2.32.0
   Compiling quote v0.6.8
   Compiling syn v0.15.11
   Compiling structopt-derive v0.2.12
    Checking structopt v0.2.12
    Checking comprakt v0.0.1 (/home/joshua/repos/by-owner/comprakt/comprakt)
    Finished dev [unoptimized + debuginfo] target(s) in 15.01s
+ cargo build
   Compiling libc v0.2.43
   Compiling unicode-width v0.1.5
   Compiling bitflags v1.0.4
   Compiling vec_map v0.8.1
   Compiling strsim v0.7.0
   Compiling ansi_term v0.11.0
   Compiling textwrap v0.10.0
   Compiling atty v0.2.11
   Compiling clap v2.32.0
   Compiling structopt v0.2.12
   Compiling comprakt v0.0.1 (/home/joshua/repos/by-owner/comprakt/comprakt)
    Finished dev [unoptimized + debuginfo] target(s) in 14.02s
+ cargo check
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
```

Reply:
ok, so only the `cargo check` is unnecessary.